### PR TITLE
Make failsafe disarm if motors armed and throttle low

### PR DESCRIPTION
**Pull request by ethomas997**
Currently, if the copter is armed, the throttle has not been raised, and radio contact is lost, a failsafe condition will occur and the motors will be commanded to the 'failsafe_throttle' value.  This is a safety issue, especially if MOTOR_STOP is enabled.  A wonky radio link could cause an unexpected spin-up of motors.  Even if MOTOR_STOP is disabled, there could be a significant increase in motor speed.

What should happen is that if the motors have not been "spun up," the copter should be disarmed.  (Other flight controllers operate this way.)  The easy / lazy way to make this happen would be to the have the failsafe code disarm if the throttle is down (low).  But there could be situations, like long descents or 3D flying, where this could be problematic.

So, what I've implemented is to have the code track the commanded state of the motors.  If the copter is armed and, during the previous three seconds, the motors have been consistently at or below the configured 'min_throttle' value and radio contact is lost, the copter is disarmed.  If the same condition happens except that the motors have been above 'min_throttle', the failsafe "landing" procedure happens.  This should account for situations where the copter is not flying (just disarm), while not being able to cause undesired failsafe-disarms in the air.

I've also fixed an issue with the 'failsafeIsIdle()' function where it would never show failsafe as 'idle' after it had been activated (causing failsafe beeping to continue after radio contact was restored).  The issue seemed to be that the counter was being polled after it was incremented by 'failsafeOnRxCycle()' but before it was reset by 'failsafeReset()' or 'failsafeOnValidDataReceived()'.  And, I improved some of the code documentation, as the logic in "failsafe.c" can be hard to follow.

--ET

### CODE REVIEW

**Comment by thenickdude** [comment]
This seems like it wouldn't work in conjunction with pid_at_min_throttle = 1, because the motors are very unlikely to sit at zero throttle when throttle is zero and MOTOR_STOP is not enabled, even when on the ground. However, in that situation it would end up doing the same as the behaviour as before the patch, so it wouldn't be any worse.

Also, it's not compatible with the 3D feature, where motor commands sit near deadband3d_low & deadband3d_high when armed and throttle is "zero", and go as low as mincommand when maximum upward thrust is commanded (i.e. below minthrottle).

**Comment by ethomas997** [comment]
Having 'pid_at_min_throttle' set is not a problem because the 'rcCommand[THROTTLE]' value used to check the motors' throttle is before the PID loop.  I've double-checked and this patch works as described when "pid_at_min_throttle=1" and MOTOR_STOP is disabled.

For 3D mode, it probably makes sense right now the have the failsafe functionality be unchanged, so I've added a check to explicitly do that.  I've also done a little improvement to how it operates when arming via TX switch (not allowing arming flag to be set during failsafe).

**Comment by thenickdude** [comment]
Eh? Oh in that case it's not really a motor counter at all, but rather an Rx throttle counter. Sorry, your description confused me (in fact, now that I look closer it has nothing to do with motors at all?).

**Comment by ethomas997** [comment]
If I'm reading the code right, 'rcData[THROTTLE]' is the reading from the TX throttle stick, while 'rcCommand[THROTTLE]' is the stick value after things like expo are applied.  The mixer generates motor output values using the 'rcCommand[THROTTLE]' value as a baseline and then applying the mixes.  So that's way I called it the "commanded state of the motors."  Regardless of the nomenclature, I think 'rcCommand[THROTTLE]' is the better one to use because it more closely indicates what the motors are doing.  As far as I can tell the patch is working well to its intended purpose.

--ET

**Comment by ethomas997** [comment]
If #185 "Modified failsafe (issue #112)" is what's going to be merged in, I'm alright with redoing this PR to go in after that (assuming this is wanted).

--ET

**Comment by ProDrone** [comment]
@ethomas997 I agree this is a safety issue but find it hard to think of a solid solution to cover it. I easily have several seconds at minimum throttle in flight and don't like the idea that the fail safe system thinks of that as being on the ground... A little off-topic perhaps, but i think MOTOR_STOP should not exist, i see that as a bigger risk than raised issue. I wonder why it is ever made.

**Comment by ethomas997** [comment]
A full 3+ seconds of motors at constant minimum throttle seems like a very unusual condition.  The failsafe would have to occur after the 3+ seconds for the disarm to happen (even more unlikely), and, if the copter is in that state, a failsafe-disarm may be a better action.

I would posit that a much more likely scenario is to have a copter with a wonky radio (loose wire, etc) sitting on the ground (before takeoff or after landing) when a failsafe occurs.  Having MOTOR_STOP disabled does not fix this because you still have a copter on the ground running (with motors at whatever 'min_throttle' value is configured) with no operator control.  Not a good situation.

I wouldn't think it should be controversial to say that a failsafe that occurs while a copter is on the ground should result in the copter disarming, not spinning up.  I know this is how the APM controllers operate.  If someone has a better way to do this I'm all ears, but otherwise I think I have pretty good solution that improves things.

BTW, I think there are lots of pilots (myself included) who fly with MOTOR_STOP enabled.  One reason is that if the copter tumbles after landing, the motors shut down right away when the throttle is lowered.  Lots of flyers, lots of preferences. :]

--ET

**Comment by ProDrone** [comment]
@ethomas997 Still agree with the safety aspect. Over the weekend i will implement it so it disarms when on the ground.

**Comment by ethomas997** [comment]
@ProDrone : Sounds good.  One thing I found was that once a failsafe was triggered the 'failsafeIsIdle()' function would never go back to returning 'false' and then the failsafe beeping never stops (until a restart).  Other users have also noticed this:  http://www.rcgroups.com/forums/showthread.php?t=2249574&page=370#post31237648

I'm using a serial-RX satellite receiver, so the issue may not happen with other receiver setups.  I think it's a race condition where the "failsafeState.counter" value is never at zero when it's polled.  This fix I found was to make the function like this:

    bool failsafeIsIdle(void)
    {             // <= 1 means failsafeReset/ValidData fn has been called
        return failsafeState.counter <= 1;
    }
So, while you're making changes, I'd recommend putting in this fix (or something like it).

--ET

File: src/main/flight/failsafe.c [hunk 1]
```diff
@@ -58,19 +72,23 @@ void useFailsafeConfig(failsafeConfig_t *failsafeConfigToUse)
...
     failsafeState.enabled = false;
+    failsafeState.motorsCounter = 0;
 
     return &failsafeState;
 }
 
 bool failsafeIsIdle(void)
-{
-    return failsafeState.counter == 0;
+{             // <= 1 means failsafeReset/ValidData fn has been called
```
Comment (NullVoxPopuli) [inline src/main/flight/failsafe.c]: formatting

**Comment by ProDrone** [comment]
@ethomas997 I added these modification to #185 and will test them tomorrow. If that turns out well i will push and rebase it (again...:), in the hope that it will soon be merged. Maybe you should close this pull-request.

**Comment by ethomas997** [comment]
Functionality implemented in PR #185

--ET
